# Analysis of borrowers’ defaulting risk

## Project description

Our project is to prepare a report for a bank’s loan division. We’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Our report will be considered when building a **credit scoring** of a potential customer. A **credit scoring** is used to evaluate the ability of a potential borrower to repay their loan.

## Description of the data

**children:** the number of children in the family

**days_employed:** how long the customer has been working

**dob_years:** the customer’s age

**education:** the customer’s education level

**education_id:** identifier for the customer’s education

**family_status:** the customer’s marital status

**family_status_id:** identifier for the customer’s marital status

**gender:** the customer’s gender

**income_type:** the customer’s income type

**debt:** whether the customer has ever defaulted on a loan

**total_income:** monthly income

**purpose:** reason for taking out a loan

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Open-the-data-file-and-have-a-look-at-the-general-information." data-toc-modified-id="Open-the-data-file-and-have-a-look-at-the-general-information.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Open the data file and have a look at the general information.</a></span><ul class="toc-item"><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></li><li><span><a href="#Data-preprocessing" data-toc-modified-id="Data-preprocessing-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data preprocessing</a></span><ul class="toc-item"><li><span><a href="#Defining-the-number-of-negative-values-on-'days_employed'-column" data-toc-modified-id="Defining-the-number-of-negative-values-on-'days_employed'-column-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Defining the number of negative values on 'days_employed' column</a></span></li><li><span><a href="#Analyzing-the-problem-on-'days_employed'" data-toc-modified-id="Analyzing-the-problem-on-'days_employed'-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Analyzing the problem on 'days_employed'</a></span></li><li><span><a href="#Processing-missing-values" data-toc-modified-id="Processing-missing-values-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Processing missing values</a></span><ul class="toc-item"><li><span><a href="#Determining-Outlier-Values" data-toc-modified-id="Determining-Outlier-Values-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Determining Outlier Values</a></span></li><li><span><a href="#Missing-Values" data-toc-modified-id="Missing-Values-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Missing Values</a></span></li></ul></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Conclusion</a></span></li><li><span><a href="#Data-type-replacement" data-toc-modified-id="Data-type-replacement-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Data type replacement</a></span></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Conclusion</a></span></li><li><span><a href="#Processing-duplicates" data-toc-modified-id="Processing-duplicates-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Processing duplicates</a></span><ul class="toc-item"><li><span><a href="#Looking-for-unique-values-in-'education'-column" data-toc-modified-id="Looking-for-unique-values-in-'education'-column-2.7.1"><span class="toc-item-num">2.7.1&nbsp;&nbsp;</span>Looking for unique values in 'education' column</a></span><ul class="toc-item"><li><span><a href="#Looking-for-duplicates-on-'marital_status'-column" data-toc-modified-id="Looking-for-duplicates-on-'marital_status'-column-2.7.1.1"><span class="toc-item-num">2.7.1.1&nbsp;&nbsp;</span>Looking for duplicates on 'marital_status' column</a></span></li><li><span><a href="#Looking-for-duplicates-on-'job'-column" data-toc-modified-id="Looking-for-duplicates-on-'job'-column-2.7.1.2"><span class="toc-item-num">2.7.1.2&nbsp;&nbsp;</span>Looking for duplicates on 'job' column</a></span></li><li><span><a href="#Looking-for-duplicates-on-'purpose'-column" data-toc-modified-id="Looking-for-duplicates-on-'purpose'-column-2.7.1.3"><span class="toc-item-num">2.7.1.3&nbsp;&nbsp;</span>Looking for duplicates on 'purpose' column</a></span></li><li><span><a href="#Importing-libraries-for-stemming-and-lemmatization" data-toc-modified-id="Importing-libraries-for-stemming-and-lemmatization-2.7.1.4"><span class="toc-item-num">2.7.1.4&nbsp;&nbsp;</span>Importing libraries for stemming and lemmatization</a></span></li></ul></li></ul></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Conclusion</a></span></li><li><span><a href="#Categorizing-Data" data-toc-modified-id="Categorizing-Data-2.9"><span class="toc-item-num">2.9&nbsp;&nbsp;</span>Categorizing Data</a></span><ul class="toc-item"><li><span><a href="#Total-debtors-rate" data-toc-modified-id="Total-debtors-rate-2.9.1"><span class="toc-item-num">2.9.1&nbsp;&nbsp;</span>Total debtors rate</a></span></li><li><span><a href="#General-examining-of-gender-effect" data-toc-modified-id="General-examining-of-gender-effect-2.9.2"><span class="toc-item-num">2.9.2&nbsp;&nbsp;</span>General examining of gender effect</a></span></li><li><span><a href="#General-examining-of-age-effect" data-toc-modified-id="General-examining-of-age-effect-2.9.3"><span class="toc-item-num">2.9.3&nbsp;&nbsp;</span>General examining of age effect</a></span></li><li><span><a href="#General-examining-job-effect" data-toc-modified-id="General-examining-job-effect-2.9.4"><span class="toc-item-num">2.9.4&nbsp;&nbsp;</span>General examining job effect</a></span></li><li><span><a href="#General-examining-of-having-kids-effect" data-toc-modified-id="General-examining-of-having-kids-effect-2.9.5"><span class="toc-item-num">2.9.5&nbsp;&nbsp;</span>General examining of having kids effect</a></span></li><li><span><a href="#General-examinig-of-marital-status-effect" data-toc-modified-id="General-examinig-of-marital-status-effect-2.9.6"><span class="toc-item-num">2.9.6&nbsp;&nbsp;</span>General examinig of marital status effect</a></span></li><li><span><a href="#General-examining-of-income-level-effect" data-toc-modified-id="General-examining-of-income-level-effect-2.9.7"><span class="toc-item-num">2.9.7&nbsp;&nbsp;</span>General examining of income level effect</a></span></li><li><span><a href="#Examining-of-'credit-type'-effect" data-toc-modified-id="Examining-of-'credit-type'-effect-2.9.8"><span class="toc-item-num">2.9.8&nbsp;&nbsp;</span>Examining of 'credit type' effect</a></span></li><li><span><a href="#Examining-of-'education-level'-effect" data-toc-modified-id="Examining-of-'education-level'-effect-2.9.9"><span class="toc-item-num">2.9.9&nbsp;&nbsp;</span>Examining of 'education level' effect</a></span></li></ul></li><li><span><a href="#New-Table-with-required-columns" data-toc-modified-id="New-Table-with-required-columns-2.10"><span class="toc-item-num">2.10&nbsp;&nbsp;</span>New Table with required columns</a></span><ul class="toc-item"><li><span><a href="#Marital-status-and-Kids-Effects-on-Debt-Situation" data-toc-modified-id="Marital-status-and-Kids-Effects-on-Debt-Situation-2.10.1"><span class="toc-item-num">2.10.1&nbsp;&nbsp;</span>Marital status and Kids Effects on Debt Situation</a></span></li><li><span><a href="#Maritial-Status-and-Debt-Relation" data-toc-modified-id="Maritial-Status-and-Debt-Relation-2.10.2"><span class="toc-item-num">2.10.2&nbsp;&nbsp;</span>Maritial Status and Debt Relation</a></span></li><li><span><a href="#Kids-status-and-Debt-Relation" data-toc-modified-id="Kids-status-and-Debt-Relation-2.10.3"><span class="toc-item-num">2.10.3&nbsp;&nbsp;</span>Kids status and Debt Relation</a></span></li><li><span><a href="#Income-Level-and-Debt-Relation" data-toc-modified-id="Income-Level-and-Debt-Relation-2.10.4"><span class="toc-item-num">2.10.4&nbsp;&nbsp;</span>Income Level and Debt Relation</a></span></li><li><span><a href="#Credit-Type-and-Debt-Relation" data-toc-modified-id="Credit-Type-and-Debt-Relation-2.10.5"><span class="toc-item-num">2.10.5&nbsp;&nbsp;</span>Credit Type and Debt Relation</a></span></li></ul></li></ul></li><li><span><a href="#Pivot-Tables-to-analyze-deeply" data-toc-modified-id="Pivot-Tables-to-analyze-deeply-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Pivot Tables to analyze deeply</a></span><ul class="toc-item"><li><span><a href="#Debtor-numbers-of-'marital_status'-and-'kids_status'" data-toc-modified-id="Debtor-numbers-of-'marital_status'-and-'kids_status'-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Debtor numbers of 'marital_status' and 'kids_status'</a></span></li><li><span><a href="#Total-numbers-of-'marital_status'-and-'kids_status'" data-toc-modified-id="Total-numbers-of-'marital_status'-and-'kids_status'-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Total numbers of 'marital_status' and 'kids_status'</a></span></li><li><span><a href="#Debt-rates-of-'marital_status'-and-'kids_status'" data-toc-modified-id="Debt-rates-of-'marital_status'-and-'kids_status'-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Debt rates of 'marital_status' and 'kids_status'</a></span></li><li><span><a href="#Debt-rates-of-'income_level'-and-'loan_sort'" data-toc-modified-id="Debt-rates-of-'income_level'-and-'loan_sort'-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Debt rates of 'income_level' and 'loan_sort'</a></span></li><li><span><a href="#Debt-rates-for-'marital_status'---'kids_status'-and-'income_level'----'loan_sort'" data-toc-modified-id="Debt-rates-for-'marital_status'---'kids_status'-and-'income_level'----'loan_sort'-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Debt rates for 'marital_status' - 'kids_status' and 'income_level'  - 'loan_sort'</a></span></li><li><span><a href="#Pivot-Tables-Conclusion" data-toc-modified-id="Pivot-Tables-Conclusion-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Pivot Tables Conclusion</a></span></li></ul></li><li><span><a href="#Answering-the-questions" data-toc-modified-id="Answering-the-questions-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Answering the questions</a></span><ul class="toc-item"><li><span><a href="#Is-there-a-relation-between-having-kids-and-repaying-a-loan-on-time?" data-toc-modified-id="Is-there-a-relation-between-having-kids-and-repaying-a-loan-on-time?-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Is there a relation between having kids and repaying a loan on time?</a></span><ul class="toc-item"><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></li><li><span><a href="#Is-there-a-relation-between-marital-status-and-repaying-a-loan-on-time?" data-toc-modified-id="Is-there-a-relation-between-marital-status-and-repaying-a-loan-on-time?-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Is there a relation between marital status and repaying a loan on time?</a></span><ul class="toc-item"><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></li><li><span><a href="#Is-there-a-relation-between-income-level-and-repaying-a-loan-on-time?" data-toc-modified-id="Is-there-a-relation-between-income-level-and-repaying-a-loan-on-time?-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Is there a relation between income level and repaying a loan on time?</a></span><ul class="toc-item"><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></li><li><span><a href="#How-do-different-loan-purposes-affect-on-time-repayment-of-the-loan?" data-toc-modified-id="How-do-different-loan-purposes-affect-on-time-repayment-of-the-loan?-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>How do different loan purposes affect on-time repayment of the loan?</a></span><ul class="toc-item"><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>Conclusion</a></span></li></ul></li></ul></li><li><span><a href="#General-conclusion" data-toc-modified-id="General-conclusion-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>General conclusion</a></span><ul class="toc-item"><li><span><a href="#Pivot-Tables-Summary" data-toc-modified-id="Pivot-Tables-Summary-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Pivot Tables Summary</a></span></li><li><span><a href="#General-Summary" data-toc-modified-id="General-Summary-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>General Summary</a></span></li></ul></li><li><span><a href="#Project-Readiness-Checklist" data-toc-modified-id="Project-Readiness-Checklist-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Project Readiness Checklist</a></span></li></ul></div>

## Open the data file and have a look at the general information. 

We are starting to import pandas library and read our csv file

In [1]:
import pandas as pd
from IPython.display import display
data = pd.read_csv('/datasets/credit_scoring_eng.csv')
display(data.describe())
data.info()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Conclusion

1- There are about 2200 (10%) missing values on the 'days_employed' and 'total_income' columns and number of them are same. They might belong to unemployed persons, but it should be checked with 'income_type' column. Although it seems small percentage, it can change our result if all of them are in debt.

2- Some column names can be changed, like 'dob_years' as 'age' and 'income_type' as 'job'.

3- The 'days_employed' column have got negative numbers. What is the meaning of them, we will look that. If it doesn't show us time interval since quit job, it can be changed positive. Probably there was a technical problem on this column. The values of days can be changed to years to easily understand this column.

4- The 'education' column has the same names beginning with capital letters. The duplicated values will be changed.

5- There are many duplicated values on 'purpose' column. We are going to look at this column and check with unique method.


In [2]:
data.head()

children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   

                   purpose  
0    purchase of the house  
1             car purchase  
2    purchase of the house  
3  supplementary education  
4        to have a wedding

In [3]:
data.tail()

children  days_employed  dob_years            education  education_id  \
21520         1   -4529.316663         43  secondary education             1   
21521         0  343937.404131         67  secondary education             1   
21522         1   -2113.346888         38  secondary education             1   
21523         3   -3112.481705         38  secondary education             1   
21524         2   -1984.507589         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income               purpose  
21520     35966.698  housing transactions  
21521     24959.969     purchase of a car  
21522     14347.610              property  
21523     39054.888     buying my own car  
21524     13127.587          to buy a car

In [4]:
data.sample(10)

children  days_employed  dob_years            education  education_id  \
12171         0  354289.226689         61  secondary education             1   
18128         0   -3767.248842         36    BACHELOR'S DEGREE             0   
17065         0    -118.574426         41  secondary education             1   
2729          0  332519.742268         59  secondary education             1   
9043          0  355436.354341         62  secondary education             1   
17547         0  375933.745908         53  secondary education             1   
8156          0    -718.575473         57  secondary education             1   
10421         1   -2084.330015         44  secondary education             1   
11217         2   -1006.565216         32  secondary education             1   
14688         1    -148.516486         48  secondary education             1   

         family_status  family_status_id gender income_type  debt  \
12171          married                 0      M     retiree     0   
18128          married                 0      F    employee     0   
17065        unmarried                 4      M    employee     0   
2729          divorced                 3      F     retiree     0   
9043   widow / widower                 2      F     retiree     0   
17547          married                 0      F     retiree     0   
8156           married                 0      M    employee     0   
10421          married                 0      M    employee     0   
11217          married                 0      M    business     1   
14688          married                 0      F    employee     0   

       total_income                                   purpose  
12171     19700.256  transactions with commercial real estate  
18128     14197.702                buy commercial real estate  
17065     21829.342                     purchase of the house  
2729      36700.570                                 education  
9043      10417.462                  real estate transactions  
17547     14799.563                  second-hand car purchase  
8156      33515.207                              to own a car  
10421     34545.142                                   housing  
11217     30303.284                      university education  
14688      8025.727                  getting higher education

## Data preprocessing

###### Changing Column Names

We will change columns names like this:

'children' --- 'kids'

'dob_years' --- 'age'

'family_status'  --- 'marital_status'

'family_status_id'  --- 'marital_status_id'

'income_type' --- 'job'

'total_income' --- 'income'

In [5]:
data.set_axis(['kids', 'days_employed', 'age', 'education', 'education_id', 'marital_status', 'marital_status_id', 
               'gender','job', 'debt', 'income', 'purpose'], axis='columns', inplace=True) 
display(data.head(3))

kids  days_employed  age            education  education_id marital_status  \
0     1   -8437.673028   42    bachelor's degree             0        married   
1     1   -4024.803754   36  secondary education             1        married   
2     0   -5623.422610   33  Secondary Education             1        married   

   marital_status_id gender       job  debt     income                purpose  
0                  0      F  employee     0  40620.102  purchase of the house  
1                  0      F  employee     0  17932.802           car purchase  
2                  0      M  employee     0  23341.752  purchase of the house

### Defining the number of negative values on 'days_employed' column

In [6]:
display(data[data['days_employed'] < 0]['days_employed'].count())

15906

### Analyzing the problem on 'days_employed'

Let's look at the row with minimum value of this column.

In [7]:
display(data[data['days_employed'] == data['days_employed'].min()])

kids  days_employed  age            education  education_id  \
16335     1  -18388.949901   61  secondary education             1   

      marital_status  marital_status_id gender       job  debt     income  \
16335        married                  0      F  employee     0  29788.629   

                        purpose  
16335  real estate transactions

It seems like negative values are not related to job or income, because the woman in this row is an employee and 61 old and has about 30K income.

Negative values on 'days_employed' column are not related to job status or not shows job quit day, 
so we can change them to positive with abs() method. Negative values were considered a technical problem.

In [8]:
data.days_employed = data.days_employed.abs()

We are going to add a new column which is named 'years_employed', because there are big numbers on 'days_employed' column and to see big picture is hard with those numbers. So '1' year contains '365.25' days, we can divide 'days_employed' by '365.25'

In [9]:
data['years_employed'] = data['days_employed'] / 365.25
display(data.head())

kids  days_employed  age            education  education_id  \
0     1    8437.673028   42    bachelor's degree             0   
1     1    4024.803754   36  secondary education             1   
2     0    5623.422610   33  Secondary Education             1   
3     3    4124.747207   32  secondary education             1   
4     0  340266.072047   53  secondary education             1   

      marital_status  marital_status_id gender       job  debt     income  \
0            married                  0      F  employee     0  40620.102   
1            married                  0      F  employee     0  17932.802   
2            married                  0      M  employee     0  23341.752   
3            married                  0      M  employee     0  42820.568   
4  civil partnership                  1      F   retiree     0  25378.572   

                   purpose  years_employed  
0    purchase of the house       23.101090  
1             car purchase       11.019312  
2    purchase of the house       15.396092  
3  supplementary education       11.292942  
4        to have a wedding      931.597733

In [10]:
display(data[data.duplicated(keep=False)])

kids  days_employed  age            education  education_id  \
120       0            NaN   46  secondary education             1   
520       0            NaN   35  secondary education             1   
541       0            NaN   57  secondary education             1   
554       0            NaN   60  secondary education             1   
680       1            NaN   30    bachelor's degree             0   
...     ...            ...  ...                  ...           ...   
20702     0            NaN   64  secondary education             1   
21032     0            NaN   60  secondary education             1   
21132     0            NaN   47  secondary education             1   
21281     1            NaN   30    bachelor's degree             0   
21415     0            NaN   54  secondary education             1   

          marital_status  marital_status_id gender            job  debt  \
120              married                  0      F       employee     0   
520    civil partnership                  1      F       employee     0   
541              married                  0      F       employee     0   
554              married                  0      M       employee     0   
680              married                  0      F  civil servant     0   
...                  ...                ...    ...            ...   ...   
20702            married                  0      F        retiree     0   
21032            married                  0      F        retiree     0   
21132            married                  0      F       employee     0   
21281            married                  0      F       employee     0   
21415            married                  0      F        retiree     0   

       income                              purpose  years_employed  
120       NaN                 university education             NaN  
520       NaN                    to have a wedding             NaN  
541       NaN             second-hand car purchase             NaN  
554       NaN                      buy real estate             NaN  
680       NaN  purchase of the house for my family             NaN  
...       ...                                  ...             ...  
20702     NaN              supplementary education             NaN  
21032     NaN                   to become educated             NaN  
21132     NaN                   housing renovation             NaN  
21281     NaN           buy commercial real estate             NaN  
21415     NaN                 housing transactions             NaN  

[106 rows x 13 columns]

### Processing missing values

#### Determining Outlier Values

In [11]:
display(data['years_employed'].max())

1099.9463394259549

What a big number. Who has been employed for almost 1100 year? There are unrealistic values on 'years_employed' and also 'days_employed' column.

To see outliners, we must determine limit value. Above that, numbers will be defined outliners. We can set '65' number to clearly determine how many outliner values are in 'years_employed'. If we assign it to 'possible_max_year' variable, we can change whatever we need that.

In [12]:
possible_max_year = 65
display(data[data['years_employed'] > possible_max_year]['years_employed'].count())

3445

There are almost 3500 outlier values, it is 16% of data. The cause of the negative value errors also can be different.
But we don't want to drop this amount. It can significantly affect our results. We can simulate average people and we must use an average value to change outliners. We cannot rely mean value, because there are a lot of missing values and outliners. So median should be used to fill NaN values.

But first, are the missing values MAR (Missing at Randomly)?

#### Missing Values

It seems like missing values on 'days_employed' and 'income' are on the same rows, because there are same NaN value quantity on these columns. We can compare this column to 'income' column.

We can use isna() and sum() methods to find total numbers of NaN values on these columns.

In [13]:
display(data['income'].isna().sum())
display(data['days_employed'].isna().sum())

2174

2174

They have same numbers (2174). We need to fill missing values with '0' to compare and determine whether all of them are belong to same rows. We can use fillna() method. After that we can use isna() method with again sum() method to check whether there are not NaN values.

In [14]:
data['days_employed'].fillna(0, inplace=True)
data['income'].fillna(0, inplace=True)

display(data['income'].isna().sum())
display(data['days_employed'].isna().sum())

0

0

To make calculation and to get rid of float numbers, we can change column types to int with astype() method.

In [15]:
data['days_employed'] = data['days_employed'].astype(int)
data['income'] = data['income'].astype(int)
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
kids                 21525 non-null int64
days_employed        21525 non-null int64
age                  21525 non-null int64
education            21525 non-null object
education_id         21525 non-null int64
marital_status       21525 non-null object
marital_status_id    21525 non-null int64
gender               21525 non-null object
job                  21525 non-null object
debt                 21525 non-null int64
income               21525 non-null int64
purpose              21525 non-null object
years_employed       19351 non-null float64
dtypes: float64(1), int64(7), object(5)
memory usage: 2.1+ MB


In how many rows, 'days_employed' and 'income' values are equal to '0' and equal to each other?
It will demonstrate us that these values are on the same rows.
To look for that, we are going to use 'for loop' with a conditional compare. It returns the number of counts of the '0' on 'days_employed' and 'income' columns on the same rows.

In [16]:
same_row_counts = 0
for row in range(len(data)):
    days_employed = data.loc[row, 'days_employed']
    income = data.loc[row, 'income']
    if days_employed == 0:
        if income == 0:
            same_row_counts += 1
display(same_row_counts)

2174

The number is same with previous isna().count() value. ('2174')
The missing values on the 'days_employed' and 'income' columns belong to same persons.
It means these missing values are  MNAR (Missing not at Random).
We will keep the values and replace '0' with mean or median value.
To determine which should be used, we had looked for outliners on 'years_employed' column. We should use median value.

We can continue with 'years_employed' column, it is the year state of 'days_column'.
median() method is used to find the median. To use mean() mislead us.

In [17]:
years_employed_median = data['years_employed'].median()
display(years_employed_median)

6.007448506170281

It is an acceptable number of average employed year and it can be used instead of missing values and outliners values.

We are going to replace missing values with median. We will use fillna() method to fill and isna() and sum() method to check them .

In [18]:
data['years_employed'] = data['years_employed'].fillna(years_employed_median)
display(data['years_employed'].isna().sum())

0

Missing values on 'years_employed' column were filled, but outliners are still in the table. It had been considered 65  as a maximum value for this column and it was in 'possible_max_year' variable. We will place it to outliner's places.

In [19]:
data.loc[data['years_employed'] > possible_max_year, 'years_employed'] = years_employed_median

'years_employed' column can be changed to int() with astype(int) method.

In [20]:
data['years_employed'] = data['years_employed'].astype(int)

'value_counts' method is used to see kids values

In [21]:
display(data['kids'].value_counts())

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: kids, dtype: int64

There are 76 people who have 20 kids. In my opinion, there could be typos in this column. There are't any number between 5 and 20. It shows us anormality of this number. Probably it should be '2' in place of '20'. It is almost 4% of other '2055' person and it doesn't affect our data, so '20' can be replaced with number '2'. Also number '-1' is probably accidentally written and '47' quantity is almost 1% of 4818. It can be changed with number 1.

In [22]:
data['kids'] = data['kids'].replace(20, 2)
data['kids'] = data['kids'].replace(-1, 1)
display(data['kids'].value_counts())

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: kids, dtype: int64

It seems more meaningful now.

We are going to examine 'age' column now.

In [23]:
display(data['age'].value_counts())

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: age, dtype: int64

The '0' counts 101. There are '101' unborn people. 

The rows which have '0' age will be removed from dataset. They can affect when they categorized by age and there could be inconsistency with 'years_employed' column. In my opinion, it will not have a significant impact on all data, but can be significantly misleading when ranked by age group. We can use drop() method to remove these rows.

In [24]:
data.drop(index=data.loc[data.loc[:, 'age'] == 0, 'age'].index, inplace=True)
display(data.tail())
display(data['age'].min())

kids  days_employed  age            education  education_id  \
21520     1           4529   43  secondary education             1   
21521     0         343937   67  secondary education             1   
21522     1           2113   38  secondary education             1   
21523     3           3112   38  secondary education             1   
21524     2           1984   40  secondary education             1   

          marital_status  marital_status_id gender       job  debt  income  \
21520  civil partnership                  1      F  business     0   35966   
21521            married                  0      F   retiree     0   24959   
21522  civil partnership                  1      M  employee     1   14347   
21523            married                  0      M  employee     1   39054   
21524            married                  0      F  employee     0   13127   

                    purpose  years_employed  
21520  housing transactions              12  
21521     purchase of a car               6  
21522              property               5  
21523     buying my own car               8  
21524          to buy a car               5

19

We should rearrange indicies with reset_index(), because last index shows still first data's length.

In [25]:
data.reset_index(drop=True, inplace=True)
data.tail()

kids  days_employed  age            education  education_id  \
21419     1           4529   43  secondary education             1   
21420     0         343937   67  secondary education             1   
21421     1           2113   38  secondary education             1   
21422     3           3112   38  secondary education             1   
21423     2           1984   40  secondary education             1   

          marital_status  marital_status_id gender       job  debt  income  \
21419  civil partnership                  1      F  business     0   35966   
21420            married                  0      F   retiree     0   24959   
21421  civil partnership                  1      M  employee     1   14347   
21422            married                  0      M  employee     1   39054   
21423            married                  0      F  employee     0   13127   

                    purpose  years_employed  
21419  housing transactions              12  
21420     purchase of a car               6  
21421              property               5  
21422     buying my own car               8  
21423          to buy a car               5

We will look at 'income' column.

In [26]:
display(data['income'].describe())

count     21424.000000
mean      24087.530760
std       17600.120155
min           0.000000
25%       14180.250000
50%       21683.500000
75%       31281.000000
max      362496.000000
Name: income, dtype: float64

There are missing values on this column. We are going to fill with median values in place of them. There is not any significant difference between mean and median, so we can use also median. It shows us also there are so many people have high income in the table. Although there are almost 2200 missing values, mean and median values are close and too lower than maximum value. The other option, there are some outliners who has so high income, but it is unknown and unmeasurable. How can we know who earns how much? For that reason, we can consider that no outlier is on this column.

In [27]:
income_median = int(data['income'].median())
display(income_median)

21683

Let's replace '0' values with median. But before we should take a copy, because after this line it gives a copy() error, when we didn't take a copy. Copied data name will be 'bank_data' and we use it after this step.

In [28]:
bank_data = data.copy()
bank_data.loc[bank_data['income'] == 0, 'income'] = income_median
display(bank_data['income'].describe())

count     21424.000000
mean      26277.691888
std       15714.407658
min        3306.000000
25%       17251.500000
50%       21683.500000
75%       31281.000000
max      362496.000000
Name: income, dtype: float64

The zeros aren't here, minimum valeu is 3306 now.

We are going to look at the 'gender' column.

In [29]:
display(bank_data['gender'].value_counts())

F      14164
M       7259
XNA        1
Name: gender, dtype: int64

There is '1' unidintified gender type and we are going to delete this row with drop() method. We can consider that this is a missing value. 

In [30]:
bank_data.drop(index=bank_data.loc[bank_data.loc[:, 'gender'] == 'XNA', 'gender'].index, inplace=True)
bank_data.reset_index(inplace=True)
display(bank_data['gender'].value_counts())

F    14164
M     7259
Name: gender, dtype: int64

Last control, 'debt' column

In [31]:
display(bank_data['debt'].unique())

array([0, 1])

There are only 0 and 1 as it's supposed to be.

### Conclusion

1_ We found missing values on 'days_employed' and 'income'. They were in the same rows. We thought there was a technical problem because there were negative values, missing values and huge values and we decided to keep these rows. We used the median in place of none values to simulate average persons. We also changed the types of this column to integer.

2_ We added the 'years_employed' named new column to realize the numbers on the 'days_employed' column and we recognized the outliners. There were huge numbers and we also changed them with median. We determined '65' years as a maximum employed year. It was even so a big number to be employed, but it was maximum of possible year front my viewpoint. Think of that 'An 80-year uncle comes and answer to 'How long had you been employed?'-'I have been working since 15 years-old.' What a young man, God give you a long life! Meanwhile, we had got a backup with dropped missing values named 'data_dropped'.

3_ We looked on 'kids' column and detected some typos and added them to related numbers.

4_ We checked 'age' column and there were '0' also. We couldn't assign to that median or mean. If we want to categorize this column, it would be effect on our result. But totaly 101 rows have no effects in total data. We dropped these rows.

5_ There was '1' missing value on the 'gender' column and we dropped it.

### Data type replacement

We had changed the float types of 'days_employed', 'income' and 'years_employed' columns to integer.

### Conclusion

There were many numbers after dot and they had no effect for big picture. We got rid off them and convert to integer. They seem nice now.

### Processing duplicates

#### Looking for unique values in 'education' column

In [32]:
display(bank_data['education'].unique())

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

All of the duplicates have different capitalizing problem.

We can solve this problem with Series.str.lower() method and then check with unique() method.

In [33]:
bank_data.loc[:, 'education'] = bank_data.loc[:, 'education'].str.lower()
display(bank_data['education'].unique())

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

All of the column is now lower case

'groupby()' and 'value_counts()' were used to check whether 'education_id' and 'education' are consistent.

In [34]:
bank_data.groupby('education_id')['education'].value_counts()

education_id  education          
0             bachelor's degree       5225
1             secondary education    15169
2             some college             741
3             primary education        282
4             graduate degree            6
Name: education, dtype: int64

It would be better if we make the training sequence logical. 'education_id' will be sorted from '0' to '4', from lowest to highest education level.

In [35]:
def education_id_define(row):
    
    education = row['education']
    
    if education == 'graduate degree':
        return 4
    
    elif education == "bachelor's degree":
        return 3
    
    elif education == 'some college':
        return 2
    
    elif education == 'secondary education':
        return 1
    
    else:
        return 0

bank_data['education_id'] = bank_data.apply(education_id_define, axis=1)
display(bank_data.groupby('education_id')['education'].value_counts())

education_id  education          
0             primary education        282
1             secondary education    15169
2             some college             741
3             bachelor's degree       5225
4             graduate degree            6
Name: education, dtype: int64

##### Looking for duplicates on 'marital_status' column

Let's check 'marital_status' and accordance 'marital_status' with 'marital_satus_id'.

In [36]:
bank_data.groupby('marital_status_id')['marital_status'].value_counts()

marital_status_id  marital_status   
0                  married              12331
1                  civil partnership     4155
2                  widow / widower        955
3                  divorced              1185
4                  unmarried             2797
Name: marital_status, dtype: int64

All unique, that's good.

##### Looking for duplicates on 'job' column

Another unique words check on 'job' column

In [37]:
display(bank_data['job'].unique())

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

All unique, there aren't duplicates

##### Looking for duplicates on 'purpose' column

Let's look at 'purpose' column. It seems like it will be hard.

In [38]:
display(bank_data['purpose'].unique())

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

I think the reason of chaos is type of data entry. In my opinion, people wrote this column themselves. If we offered multiple choices, both their and our work would be easier. Maybe we can offer this idea to whom is responsible to get data. It would be better.

#####  Importing libraries for stemming and lemmatization

To analyse 'purpose' column, we need to import NLTK (Natural Language Toolkit) library and some methods from there.

In [39]:
import nltk 
from nltk.stem import SnowballStemmer
english_stemmer = SnowballStemmer('english')
from nltk.stem import WordNetLemmatizer
wordnet_lemma = WordNetLemmatizer()
from nltk.tokenize import word_tokenize

First let's look which words there are

In [40]:
unique_list = list(bank_data['purpose'].unique())
words = []
for text in unique_list:
    text_list = text.split()
    for w in text_list:
        if w not in words:
            words.append(w)
print(words)

['purchase', 'of', 'the', 'house', 'car', 'supplementary', 'education', 'to', 'have', 'a', 'wedding', 'housing', 'transactions', 'having', 'for', 'my', 'family', 'buy', 'real', 'estate', 'commercial', 'residential', 'construction', 'own', 'property', 'building', 'buying', 'second-hand', 'with', 'cars', 'become', 'educated', 'getting', 'an', 'ceremony', 'get', 'higher', 'profile', 'university', 'renting', 'out', 'renovation', 'going']


After that let's find out the stems of the words

In [41]:
stems = []
for s in words:
    st = english_stemmer.stem(s)
    stems.append(st)
print(stems)

['purchas', 'of', 'the', 'hous', 'car', 'supplementari', 'educ', 'to', 'have', 'a', 'wed', 'hous', 'transact', 'have', 'for', 'my', 'famili', 'buy', 'real', 'estat', 'commerci', 'residenti', 'construct', 'own', 'properti', 'build', 'buy', 'second-hand', 'with', 'car', 'becom', 'educ', 'get', 'an', 'ceremoni', 'get', 'higher', 'profil', 'univers', 'rent', 'out', 'renov', 'go']


Last, finding lemmas of the words

In [42]:
lemmas = []
for l in words:
    lm = wordnet_lemma.lemmatize(l)
    lemmas.append(lm)
print(lemmas)

['purchase', 'of', 'the', 'house', 'car', 'supplementary', 'education', 'to', 'have', 'a', 'wedding', 'housing', 'transaction', 'having', 'for', 'my', 'family', 'buy', 'real', 'estate', 'commercial', 'residential', 'construction', 'own', 'property', 'building', 'buying', 'second-hand', 'with', 'car', 'become', 'educated', 'getting', 'an', 'ceremony', 'get', 'higher', 'profile', 'university', 'renting', 'out', 'renovation', 'going']


We are going to run a loop. It will categorize by names.

In [43]:
credit_type_list = []
purpose_list = bank_data.loc[:, 'purpose']

for rw_text in purpose_list:
    if 'hous' in rw_text:
        credit_type_list.append('house')
        continue
    if 'car' in rw_text:
        credit_type_list.append('car')
        continue
    if 'educ' in rw_text:
        credit_type_list.append('education')
        continue
    if 'wedding' in rw_text:
        credit_type_list.append('wedding')
        continue
    if 'residenti' in rw_text:
        credit_type_list.append('house')
        continue
    if 'estate' in rw_text:            
        credit_type_list.append('estate')
        continue
        
    if 'univers' in rw_text:
        credit_type_list.append('education')
        continue
        
    if 'properti' or 'property' in rw_text:
        credit_type_list.append('property')
        continue

    else:
        credit_type_list.append('other')
        continue
        
display(len(credit_type_list))
bank_data['credit_type'] = credit_type_list
display(bank_data['credit_type'].unique())

21423

array(['house', 'car', 'education', 'wedding', 'estate', 'property'],
      dtype=object)

Let's check compatibility of the 'credit_type' and 'purpose' with groupby and value_counts()

In [44]:
bank_data.groupby('credit_type')['purpose'].value_counts()

credit_type  purpose                                 
car          buying my own car                           502
             car                                         491
             second-hand car purchase                    487
             buying a second-hand car                    479
             to own a car                                477
             cars                                        476
             to buy a car                                467
             car purchase                                460
             purchase of a car                           454
education    going to university                         493
             supplementary education                     460
             university education                        450
             to get a supplementary education            445
             education                                   444
             getting an education                        442
             profile education 

It seems nice.

In [45]:
display(bank_data.head())

index  kids  days_employed  age            education  education_id  \
0      0     1           8437   42    bachelor's degree             3   
1      1     1           4024   36  secondary education             1   
2      2     0           5623   33  secondary education             1   
3      3     3           4124   32  secondary education             1   
4      4     0         340266   53  secondary education             1   

      marital_status  marital_status_id gender       job  debt  income  \
0            married                  0      F  employee     0   40620   
1            married                  0      F  employee     0   17932   
2            married                  0      M  employee     0   23341   
3            married                  0      M  employee     0   42820   
4  civil partnership                  1      F   retiree     0   25378   

                   purpose  years_employed credit_type  
0    purchase of the house              23       house  
1             car purchase              11         car  
2    purchase of the house              15       house  
3  supplementary education              11   education  
4        to have a wedding               6     wedding

### Conclusion

1- First, we solved duplicated value issues on 'education' column. There was capitalizing problem and we arranged them.

2- Then we focused 'purpose' column, we found all purposes and get a list from them. After that we got stems and lemmas. We picked up some of them to create 'credit_type'. We used for loop to get new list with a conditional situation of words in our list. We compared words with stems or lemmas. We categorized words by a condition. We added a new column named 'credit_type'.

### Categorizing Data

We can categorize our table for many criteria. The debt status could be affected by age, employment, gender, income, kids number and marital status. How can we determine which parameter how much effect our data? We should think and answer them. Maybe we can reduce parameters. Our goal is to find effect of having kids, marital status and income variables. We will also look for which credit_type effects loan repayment on time. Our result criteria is debt status. We can find total numbers for every parameters which are in debt, which aren't? And then we can compare the results dividing each of them by total of each and find a percentage rate. Let's start.

First of all, we need to simplify our table. We are creating a new table named 'credit_data'. We can change the order of columns.

In [46]:
credit_data = bank_data.loc[:, ['gender', 'age', 'job', 'years_employed', 'education', 'education_id', 'marital_status', 'marital_status_id',
                       'kids', 'credit_type', 'income', 'debt']]
display(credit_data.head())

gender  age       job  years_employed            education  education_id  \
0      F   42  employee              23    bachelor's degree             3   
1      F   36  employee              11  secondary education             1   
2      M   33  employee              15  secondary education             1   
3      M   32  employee              11  secondary education             1   
4      F   53   retiree               6  secondary education             1   

      marital_status  marital_status_id  kids credit_type  income  debt  
0            married                  0     1       house   40620     0  
1            married                  0     1         car   17932     0  
2            married                  0     0       house   23341     0  
3            married                  0     3   education   42820     0  
4  civil partnership                  1     0     wedding   25378     0

It looks nicer than before.

#### Total debtors rate

We can find our rates. We will look for some variables effect, after that we'll look for main goal. Let's start with total debtors rate. 

In [47]:
display(credit_data['debt'].describe())

count    21423.000000
mean         0.080894
std          0.272679
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: debt, dtype: float64

Let's print our important values.

In [48]:
non_debtors = credit_data[credit_data['debt'] == 0]['debt'].count()   # total number of people who have no debt
debtors = credit_data[credit_data['debt'] == 1]['debt'].count()  # total number of people who have debt
total_people = debtors + non_debtors  # total people in the tale
total_debtor_rate = debtors / total_people  # debtors percentage
print("{} people are without debt\n{} people are in debt\nTotal debtor rate is {:.1%}".format(non_debtors, debtors, total_debtor_rate))

19690 people are without debt
1733 people are in debt
Total debtor rate is 8.1%


8 out of every 100 people are in debt.

We are now going to some general conclusions for main variables.

#### General examining of gender effect

Let's examine first 'gender' column. We will group table by 'gender' and look 'debt' values and then examine with describe() method.

In [49]:
gender_debt = credit_data.groupby('gender')['debt']
display(gender_debt.describe())

count      mean       std  min  25%  50%  75%  max
gender                                                      
F       14164.0  0.069825  0.254861  0.0  0.0  0.0  0.0  1.0
M        7259.0  0.102493  0.303317  0.0  0.0  0.0  0.0  1.0

Females are almost twice of males in our data.

Kudos to females. Females' debt rate is 1% below the total debt rate and males' 3% above. Although males are half of females, male debtors are almost as much as female debtors. Despite the 3% figure from the average seems small, half of the people we will examine debt status are male, according to this situation, 'gender' seems to be the most important variable in determining the credit score.

#### General examining of age effect

There are a lot of 'age' values in the table. If we categorize 'age' values, we can read age effect clearly. We will define a function to group age data and add a new column with that. But before, we are going to look for min, max, and average ages in the table.

In [50]:
display(credit_data['age'].describe())

count    21423.000000
mean        43.498390
std         12.246495
min         19.000000
25%         33.000000
50%         43.000000
75%         53.000000
max         75.000000
Name: age, dtype: float64

There aren't any children. Equality of mean and median values shows us that 'age' column has uniform and homogeneous values.

We will create a new column 'age_group' with criteria like this: 

18-39: 'young' / 40-65: 'middle_age' / 65 and up: 'elderly'

In [51]:
def age_group(row):
    age = row['age']
    if 18 <= age < 40:
        return 'young'
    elif 40 <= age < 65:
        return 'middle_age'
    else:
        return 'elderly'

credit_data['age_group'] = credit_data.apply(age_group, axis=1)
display(credit_data.head())

gender  age       job  years_employed            education  education_id  \
0      F   42  employee              23    bachelor's degree             3   
1      F   36  employee              11  secondary education             1   
2      M   33  employee              15  secondary education             1   
3      M   32  employee              11  secondary education             1   
4      F   53   retiree               6  secondary education             1   

      marital_status  marital_status_id  kids credit_type  income  debt  \
0            married                  0     1       house   40620     0   
1            married                  0     1         car   17932     0   
2            married                  0     0       house   23341     0   
3            married                  0     3   education   42820     0   
4  civil partnership                  1     0     wedding   25378     0   

    age_group  
0  middle_age  
1       young  
2       young  
3       young  
4  middle_age

Now we will group by 'age_group' and see 'debt' amounts.

In [52]:
age_group_debt = credit_data.groupby('age_group')['debt']
display(age_group_debt.describe())

count      mean       std  min  25%  50%  75%  max
age_group                                                       
elderly       899.0  0.054505  0.227138  0.0  0.0  0.0  0.0  1.0
middle_age  11668.0  0.067107  0.250217  0.0  0.0  0.0  0.0  1.0
young        8856.0  0.101739  0.302322  0.0  0.0  0.0  0.0  1.0

Old people seem to pay their debts early. But loan demands come from middle age and young people. Young people are fewer than middle age people but there are more young people in debt. While 2 out of 20 young people are in debt, 1 in 20 older people is in debt. Almost twice the elderly debt rate. "Young people attention! You are entering the risk pot for credit score."

We can find if we want to wonder which ages are mostly in debt, but it doen't necessary.

#### General examining job effect

Let's look job effect

In [53]:
job_debt = credit_data.groupby('job')['debt']
display(job_debt.describe())

count      mean       std  min   25%  50%  \
job                                                                        
business                      5064.0  0.074052  0.261882  0.0  0.00  0.0   
civil servant                 1453.0  0.059188  0.236057  0.0  0.00  0.0   
employee                     11064.0  0.095354  0.293717  0.0  0.00  0.0   
entrepreneur                     2.0  0.000000  0.000000  0.0  0.00  0.0   
paternity / maternity leave      1.0  1.000000       NaN  1.0  1.00  1.0   
retiree                       3836.0  0.056048  0.230044  0.0  0.00  0.0   
student                          1.0  0.000000       NaN  0.0  0.00  0.0   
unemployed                       2.0  0.500000  0.707107  0.0  0.25  0.5   

                              75%  max  
job                                     
business                     0.00  1.0  
civil servant                0.00  1.0  
employee                     0.00  1.0  
entrepreneur                 0.00  0.0  
paternity / maternity leave  1.00  1.0  
retiree                      0.00  1.0  
student                      0.00  0.0  
unemployed                   0.75  1.0

It seems like employees debt rate is almost 10%, while civil servants's debt rate is about 6%. We can compare business, civil servant, employee and retiree jobs. Others are small numbers and almost not an effect in big picture. We can skip them for general information.

#### General examining of having kids effect

In [54]:
kids_debt = credit_data.groupby('kids')['debt']
display(kids_debt.describe())

count      mean       std  min  25%  50%  75%  max
kids                                                      
0     14079.0  0.075147  0.263638  0.0  0.0  0.0  0.0  1.0
1      4849.0  0.091153  0.287856  0.0  0.0  0.0  0.0  1.0
2      2117.0  0.095418  0.293861  0.0  0.0  0.0  0.0  1.0
3       328.0  0.082317  0.275267  0.0  0.0  0.0  0.0  1.0
4        41.0  0.097561  0.300406  0.0  0.0  0.0  0.0  1.0
5         9.0  0.000000  0.000000  0.0  0.0  0.0  0.0  0.0

The people which have '0' and '5' children have a debt rate lower than mean, but the others' are over average debt rate. Let's even give 5 points to 5 children, zero debt. But very few amount, only 9 persons. We can not rely only this data.

#### General examinig of marital status effect

In [55]:
marital_debt = credit_data.groupby('marital_status')['debt']
display(marital_debt.describe())

count      mean       std  min  25%  50%  75%  max
marital_status                                                         
civil partnership   4155.0  0.092900  0.290327  0.0  0.0  0.0  0.0  1.0
divorced            1185.0  0.071730  0.258149  0.0  0.0  0.0  0.0  1.0
married            12331.0  0.075176  0.263686  0.0  0.0  0.0  0.0  1.0
unmarried           2797.0  0.097605  0.296832  0.0  0.0  0.0  0.0  1.0
widow / widower      955.0  0.064921  0.246516  0.0  0.0  0.0  0.0  1.0

'civil partnership' and 'unmarried' have debt rates over average total rate.

#### General examining of income level effect

We can categorize income column. Let's look some values to decide appropriate limit values for categorization

In [56]:
display(credit_data['income'].describe())

count     21423.000000
mean      26277.395650
std       15714.714614
min        3306.000000
25%       17251.000000
50%       21683.000000
75%       31278.500000
max      362496.000000
Name: income, dtype: float64

I think 20.000 and 40.000 values can be limit for income classification. we can change it according to the income level of the country.

In [57]:
def income_level(row):
    
    income = row['income']
    
    if income <= 20000:
        return 'low'
    elif 20000 <= income < 40000:
        return 'middle'
    else:
        return 'high'
    
    
credit_data['income_level'] = credit_data.apply(income_level, axis=1)
display(credit_data.head())

gender  age       job  years_employed            education  education_id  \
0      F   42  employee              23    bachelor's degree             3   
1      F   36  employee              11  secondary education             1   
2      M   33  employee              15  secondary education             1   
3      M   32  employee              11  secondary education             1   
4      F   53   retiree               6  secondary education             1   

      marital_status  marital_status_id  kids credit_type  income  debt  \
0            married                  0     1       house   40620     0   
1            married                  0     1         car   17932     0   
2            married                  0     0       house   23341     0   
3            married                  0     3   education   42820     0   
4  civil partnership                  1     0     wedding   25378     0   

    age_group income_level  
0  middle_age         high  
1       young          low  
2       young       middle  
3       young         high  
4  middle_age       middle

There are 'high', 'middle', and 'low' income levels now. We can categorize the income level easier. Let's try general look for income level - debt relation.

In [58]:
income_level_grouped = credit_data.groupby('income_level')['debt']
display(income_level_grouped.value_counts())

income_level  debt
high          0        2606
              1         193
low           0        6728
              1         605
middle        0       10356
              1         935
Name: debt, dtype: int64

Let's examine debt rates for each income level.

In [59]:
display(income_level_grouped.describe())

count      mean       std  min  25%  50%  75%  max
income_level                                                      
high           2799.0  0.068953  0.253420  0.0  0.0  0.0  0.0  1.0
low            7333.0  0.082504  0.275149  0.0  0.0  0.0  0.0  1.0
middle        11291.0  0.082809  0.275606  0.0  0.0  0.0  0.0  1.0

*** High class has a debt rate smaller total average (it was 8.1%). This situation is shows us 'income_level' affect  total debt rate slightly.

#### Examining of 'credit type' effect

In [60]:
credit_type_grouped = credit_data.groupby('credit_type')['debt']
display(credit_type_grouped.describe())

count      mean       std  min  25%  50%  75%  max
credit_type                                                     
car          4293.0  0.093175  0.290711  0.0  0.0  0.0  0.0  1.0
education    4004.0  0.092408  0.289637  0.0  0.0  0.0  0.0  1.0
estate       3856.0  0.076504  0.265838  0.0  0.0  0.0  0.0  1.0
house        4402.0  0.067015  0.250076  0.0  0.0  0.0  0.0  1.0
property     2534.0  0.074586  0.262773  0.0  0.0  0.0  0.0  1.0
wedding      2334.0  0.078835  0.269538  0.0  0.0  0.0  0.0  1.0

*** While there are difficulties in repaying 'car' and 'education' credit types, 'estate', 'property', and especially 'house' loans are successfully repaid. 

#### Examining of 'education level' effect

In [61]:
education_grouped = credit_data.groupby('education')['debt']
display(education_grouped.describe())

count      mean       std  min  25%  50%  75%  max
education                                                                
bachelor's degree     5225.0  0.053014  0.224084  0.0  0.0  0.0  0.0  1.0
graduate degree          6.0  0.000000  0.000000  0.0  0.0  0.0  0.0  0.0
primary education      282.0  0.109929  0.313357  0.0  0.0  0.0  0.0  1.0
secondary education  15169.0  0.089459  0.285414  0.0  0.0  0.0  0.0  1.0
some college           741.0  0.091768  0.288893  0.0  0.0  0.0  0.0  1.0

We can categorize education level: 'low' and 'high'. Let's define a function and create new column named 'education_level'.

In [62]:
def education_level_define(row):
    
    ident = row['education_id']
    
    if ident < 2:
        return 'low'
    
    else:
        return 'high'
    
credit_data['education_level'] = credit_data.apply(education_level_define, axis=1)
display(credit_data.head())

gender  age       job  years_employed            education  education_id  \
0      F   42  employee              23    bachelor's degree             3   
1      F   36  employee              11  secondary education             1   
2      M   33  employee              15  secondary education             1   
3      M   32  employee              11  secondary education             1   
4      F   53   retiree               6  secondary education             1   

      marital_status  marital_status_id  kids credit_type  income  debt  \
0            married                  0     1       house   40620     0   
1            married                  0     1         car   17932     0   
2            married                  0     0       house   23341     0   
3            married                  0     3   education   42820     0   
4  civil partnership                  1     0     wedding   25378     0   

    age_group income_level education_level  
0  middle_age         high            high  
1       young          low             low  
2       young       middle             low  
3       young         high             low  
4  middle_age       middle             low

Let's examine again for education_level.

In [63]:
education_level_grouped = credit_data.groupby('education_level')['debt']
display(education_level_grouped.describe())

count      mean       std  min  25%  50%  75%  max
education_level                                                      
high              5972.0  0.057770  0.233327  0.0  0.0  0.0  0.0  1.0
low              15451.0  0.089832  0.285951  0.0  0.0  0.0  0.0  1.0

Apparently, education level is an important factor in loan repayment.

### New Table with required columns

We are going to focus our goals now. First we will create a new table from old and get only related columns. Then we will categorize each column 3 category. After that we will create again a new table with new categories.

Let's make a summary table by pulling only necessary columns from our first table.

In [64]:
new_credit_data = pd.DataFrame(credit_data, columns=['marital_status', 'marital_status_id', 'kids', 'income', 'income_level', 'credit_type', 'debt'])
display(new_credit_data)

marital_status  marital_status_id  kids  income income_level  \
0                married                  0     1   40620         high   
1                married                  0     1   17932          low   
2                married                  0     0   23341       middle   
3                married                  0     3   42820         high   
4      civil partnership                  1     0   25378       middle   
...                  ...                ...   ...     ...          ...   
21418  civil partnership                  1     1   35966       middle   
21419            married                  0     0   24959       middle   
21420  civil partnership                  1     1   14347          low   
21421            married                  0     3   39054       middle   
21422            married                  0     2   13127          low   

      credit_type  debt  
0           house     0  
1             car     0  
2           house     0  
3       education     0  
4         wedding     0  
...           ...   ...  
21418       house     0  
21419         car     0  
21420    property     1  
21421         car     1  
21422         car     0  

[21423 rows x 7 columns]

We need some more editing and get new columns, but not now.

#### Marital status and Kids Effects on Debt Situation

In [65]:
new_credit_data.groupby('marital_status_id')['kids'].value_counts()

marital_status_id  kids
0                  0       7469
                   1       3004
                   2       1575
                   3        247
                   4         29
                   5          7
1                  0       2738
                   1        997
                   2        354
                   3         56
                   4          8
                   5          2
2                  0        843
                   1         81
                   2         24
                   3          6
                   4          1
3                  0        777
                   1        315
                   2         81
                   3         11
                   4          1
4                  0       2252
                   1        452
                   2         83
                   3          8
                   4          2
Name: kids, dtype: int64

This table shows us only total numbers by each category in debt. Let's define a function to categorize marital and kids status separately and create with them new columns.

In [66]:
def marital_group(row):
    
    marital = row['marital_status_id']
    
    if marital < 2:
        return 'together'

    else:
        return 'alone'
        
        
def kids_group(row):
    
    kids = row['kids']
    
    if kids == 0:
        return 'No'
    
    else:
        return 'Yes'
    
        
new_credit_data['marital_status'] = new_credit_data.apply(marital_group, axis=1)
new_credit_data['kids_status'] = new_credit_data.apply(kids_group, axis=1)
display(new_credit_data.tail(10))

marital_status  marital_status_id  kids  income income_level  \
21413       together                  0     1   17517          low   
21414       together                  0     0   51649         high   
21415       together                  1     0   28489       middle   
21416       together                  0     0   24618       middle   
21417          alone                  3     1   18551          low   
21418       together                  1     1   35966       middle   
21419       together                  0     0   24959       middle   
21420       together                  1     1   14347          low   
21421       together                  0     3   39054       middle   
21422       together                  0     2   13127          low   

      credit_type  debt kids_status  
21413   education     1         Yes  
21414       house     0          No  
21415         car     0          No  
21416         car     0          No  
21417      estate     0         Yes  
21418       house     0         Yes  
21419         car     0          No  
21420    property     1         Yes  
21421         car     1         Yes  
21422         car     0         Yes

Now we are getting a new table with only required columns and naming 'last_credit_data'

In [67]:
last_credit_data = pd.DataFrame(new_credit_data, columns=['marital_status', 'kids_status', 'income_level', 'credit_type', 'debt'])
display(last_credit_data.head())

marital_status kids_status income_level credit_type  debt
0       together         Yes         high       house     0
1       together         Yes          low         car     0
2       together          No       middle       house     0
3       together         Yes         high   education     0
4       together          No       middle     wedding     0

We are going to categorize again by each column and see the debt rates.

#### Maritial Status and Debt Relation

 We are grouping by 'maritial_status' on 'debt' column.

In [68]:
marital_debt_data = last_credit_data.groupby('marital_status')['debt']
display(marital_debt_data.describe())

count      mean       std  min  25%  50%  75%  max
marital_status                                                      
alone            4937.0  0.085072  0.279017  0.0  0.0  0.0  0.0  1.0
together        16486.0  0.079643  0.270748  0.0  0.0  0.0  0.0  1.0

*** The debt ratio of families living together is almost the same with the average, while the debt ratio of those living alone is above the general average. It can only slightly reduce the credit score of the alone customers.

#### Kids status and Debt Relation

 We are grouping our data by 'kids_status' on 'debt' column.

In [69]:
kids_debt_data = last_credit_data.groupby('kids_status')['debt']
display(kids_debt_data.describe())

count      mean       std  min  25%  50%  75%  max
kids_status                                                      
No           14079.0  0.075147  0.263638  0.0  0.0  0.0  0.0  1.0
Yes           7344.0  0.091912  0.288921  0.0  0.0  0.0  0.0  1.0

*** There are about 14.000 people who don't have kids, it is almost 2 out of 3. Their debt rate is lower than average. On the contrary, people who have kids less reliable on repayment of loan. While their amount is about 7.000 (almost 1 out of 3), their debt rate is more than 1% of average debt rate. It is about 80 people. But if we compare it to those who do not have children, numbers increase and then it makes sense. It shows us having kids has negative effect on repayment of loan.

#### Income Level and Debt Relation

In [70]:
income_debt_data = last_credit_data.groupby('income_level')['debt']
display(income_debt_data.describe())

count      mean       std  min  25%  50%  75%  max
income_level                                                      
high           2799.0  0.068953  0.253420  0.0  0.0  0.0  0.0  1.0
low            7333.0  0.082504  0.275149  0.0  0.0  0.0  0.0  1.0
middle        11291.0  0.082809  0.275606  0.0  0.0  0.0  0.0  1.0

*** As we calculated before, high class has debt rate below average, while other classes slightly above (0.2%). In my opinion, it can be considered pointless for low or middle income level, but the credit score of high income levels can be increased.

#### Credit Type and Debt Relation

We had analyzed these values under all credit types, but now let's examine them in 3 groups.

In [71]:
real_estate = ['estate', 'house', 'property']
general = ['education', 'car', 'wedding']

def type_group(row):
    
    loan_type = row['credit_type']

    if loan_type in real_estate:
        return 'real_estate'
    
    else:
        return 'general'
    
last_credit_data['loan_sort'] = last_credit_data.apply(type_group, axis=1)
display(last_credit_data.head())

marital_status kids_status income_level credit_type  debt    loan_sort
0       together         Yes         high       house     0  real_estate
1       together         Yes          low         car     0      general
2       together          No       middle       house     0  real_estate
3       together         Yes         high   education     0      general
4       together          No       middle     wedding     0      general

In [72]:
loan_sort_debt = last_credit_data.groupby('loan_sort')['debt']
display(loan_sort_debt.describe())

count      mean       std  min  25%  50%  75%  max
loan_sort                                                        
general      10631.0  0.089738  0.285819  0.0  0.0  0.0  0.0  1.0
real_estate  10792.0  0.072183  0.258803  0.0  0.0  0.0  0.0  1.0

Real estate loans have lower debt rate, while general loans have higher.

In summary,


1- Lonely people have more difficulty with loan repayments than together.

2- Those who have children have more difficulty in repaying loans than those who do not.

3- Income levels of middle and low classes have no direct effect on loan repayment, but higher-income people appear to be more successful for repayments.

4- Loan type seems to be effective in repayment rates. The reason for this may be that the mortgage status in the real estate obliges the payment.

## Pivot Tables to analyze deeply

Now let's use pivot tables to aggregate the data.

First we will compare the debt rates of two variables, 'marital_status' and 'kids_status'. (pivot3)

Secondly we are going to look other two variables, 'income_level' and 'loan_sort'. (pivot6)

Last we will place 'marital_status' and 'kids_status' on index while 'income_level' and 'loan_sort' on columns.

### Debtor numbers of 'marital_status' and 'kids_status'

They will show debtors (pivot1) and total numbers (pivot2), and at last debt percentages (pivot3) of each category. We will use 'sum' function on 'debt' values to determine debtors of each category.

In [73]:
pivot1 = last_credit_data.pivot_table(columns=['marital_status'], index=['kids_status'], values='debt', aggfunc='sum')
display(pivot1)

marital_status  alone  together
kids_status                    
No                317       741
Yes               103       572

This table shows us only numbers, but we need debtors rates to compare.

### Total numbers of 'marital_status' and 'kids_status'

Now we are getting a new pivot table with total numbers. We will use 'count' function on 'debt' values to find total people of each category.

In [74]:
pivot2 = last_credit_data.pivot_table(columns=['marital_status'], index=['kids_status'], values='debt', aggfunc='count')
display(pivot2)

marital_status  alone  together
kids_status                    
No               3872     10207
Yes              1065      6279

### Debt rates of 'marital_status' and 'kids_status'

Last we can create a new pivot table with them. It shows how many people are in debt per 100 people for each category in the pivot table.

In [75]:
pivot3 = (pivot1 / pivot2 * 100).round(1)
display(pivot3)

marital_status  alone  together
kids_status                    
No                8.2       7.3
Yes               9.7       9.1

General debt rate was 8.1%.

Let's remember average debt rates for these categories which we calculated before.

For 'marital_status':
|'together' has '8.0%'| - |'alone' has '8.5%'|

For 'kids_status':
|'No' has '7.5%'| - |'Yes' has '9.2'|

Can we say still 'kids_status' has negative effect for repayment of loan? Yes, it affects repayment of loan negative and still have high debt rates (9.7% and 9.1%). Also we can say lonely people without kids don't have big problem about loan repayment because their debt rate (8.2%) almost same as general average (8.1%). As a result, together and childless families have highest credit score with the lowest debt rate (7.3%).

*** To sum up, 

1- 'kids_status' (to have kids) affects debt rates negatively. (We mentioned about this fact, it didn't change.)

2- 'marital_status' (to be together) affects debt rates positively if they don't have any kids. (We mentioned about lonely people have more difficulty for repayment of debt. But this pivot table shows us that the reason is actually not to be alone but to have children. If lonely people have no kids, their debt rate is almost same as general debt rate. In addition to that, together people's rate is lower than lonely people. They keep the advantage.)

### Debt rates of 'income_level' and 'loan_sort'

Now we are going to look other two variables, 'income_level' and 'loan_sort'.

In [76]:
pivot4 = last_credit_data.pivot_table(columns=['income_level'], index=['loan_sort'], values='debt', aggfunc='sum')

pivot5 = last_credit_data.pivot_table(columns=['income_level'], index=['loan_sort'], values='debt', aggfunc='count')
pivot6 = (pivot4 / pivot5 * 100).round(1)
display(pivot4)
display(pivot5)
display(pivot6)

income_level  high  low  middle
loan_sort                      
general        105  353     496
real_estate     88  252     439

income_level  high   low  middle
loan_sort                       
general       1358  3660    5613
real_estate   1441  3673    5678

income_level  high  low  middle
loan_sort                      
general        7.7  9.6     8.8
real_estate    6.1  6.9     7.7

Let's remember again debt rates of each category.

For 'income_level':
|high : 6.9%| - |middle: 8.3%| - |low : 8.3%|

For 'loan_sort':
|general : 9.0%| - |real_estate : 7.2%|

*** Summary:

1- 'loan_sort' (general type) definitely has a negative effect, because it pulls all rates up on each income class.  Additionally it's obvious that debt rates for 'real_estate' are lower for each income class, so this type has positive effect for repayment. (We mentioned about that, but we saw it's effects on each category)

2- 'income_level' for high class still has lowest debt rates despite 'loan_sort', we can say again high class has positive effect. (We emphasized that before.)

3- We should pay attention 'low' income level and 'general' loan sort intersection. Their debt rate high and it reduces credit score of low-income people while they want to credit for 'general' purposes.

### Debt rates for 'marital_status' - 'kids_status' and 'income_level'  - 'loan_sort'

Nor we are going to deep numbers. We'll place 'marital_status' and 'kids_status' on index while 'income_level' and 'loan_sort' on columns. Let's see if the results will change.

In [77]:
pivot7 = last_credit_data.pivot_table(index=['marital_status', 'kids_status'], columns=['income_level', 'loan_sort'],
                                                     values='debt', aggfunc='sum')
pivot8 = last_credit_data.pivot_table(index=['marital_status', 'kids_status'], columns=['income_level', 'loan_sort'],
                                                     values='debt', aggfunc='count')
pivot9 = (pivot7 / pivot8 * 100).round(1)
display(pivot7)
display(pivot8)
display(pivot9)

income_level                  high                 low              middle  \
loan_sort                  general real_estate general real_estate general   
marital_status kids_status                                                   
alone          No               18          19      67          40      86   
               Yes               2           6      11          19      32   
together       No               46          30     156         113     211   
               Yes              39          33     119          80     167   

income_level                            
loan_sort                  real_estate  
marital_status kids_status              
alone          No                   87  
               Yes                  33  
together       No                  185  
               Yes                 134

income_level                  high                 low              middle  \
loan_sort                  general real_estate general real_estate general   
marital_status kids_status                                                   
alone          No              212         272     617         777     865   
               Yes              38          85     151         202     254   
together       No              670         639    1870        1699    2779   
               Yes             438         445    1022         995    1715   

income_level                            
loan_sort                  real_estate  
marital_status kids_status              
alone          No                 1129  
               Yes                 335  
together       No                 2550  
               Yes                1664

income_level                  high                 low              middle  \
loan_sort                  general real_estate general real_estate general   
marital_status kids_status                                                   
alone          No              8.5         7.0    10.9         5.1     9.9   
               Yes             5.3         7.1     7.3         9.4    12.6   
together       No              6.9         4.7     8.3         6.7     7.6   
               Yes             8.9         7.4    11.6         8.0     9.7   

income_level                            
loan_sort                  real_estate  
marital_status kids_status              
alone          No                  7.7  
               Yes                 9.9  
together       No                  7.3  
               Yes                 8.1

Let's look at this table from a different perspective.

In [78]:
pivot10 = last_credit_data.pivot_table(columns=['marital_status', 'loan_sort'], index=['income_level', 'kids_status'],
                                                     values='debt', aggfunc='sum')
pivot11 = last_credit_data.pivot_table(columns=['marital_status', 'loan_sort'], index=['income_level', 'kids_status'],
                                                     values='debt', aggfunc='count')
pivot12 = (pivot10 / pivot11 * 100).round(1)
display(pivot12)

marital_status             alone             together            
loan_sort                general real_estate  general real_estate
income_level kids_status                                         
high         No              8.5         7.0      6.9         4.7
             Yes             5.3         7.1      8.9         7.4
low          No             10.9         5.1      8.3         6.7
             Yes             7.3         9.4     11.6         8.0
middle       No              9.9         7.7      7.6         7.3
             Yes            12.6         9.9      9.7         8.1

Let's consider again debt rates we calculated before:

For 'marital_status': |'together' has '8.0%'| - |'alone' has '8.5%'|

For 'kids_status': |'No' has '7.5%'| - |'Yes' has '9.2'|

For 'income_level': |high : 6.9%| - |middle: 8.3%| - |low : 8.3%|

For 'loan_sort': |general : 9.0%| - |real_estate : 7.2%|

### Pivot Tables Conclusion

If we need to evaluate according to the average debt ratio (8.1%)

1- Those who are in the 'middle' income and have kids have generally over average debt rates. They are the most risky group. (middle-with kids)

2- In 'middle' income level - for 'general' loan type, lonely people who have kids have highest debt rate. (12.6%)

2- 'alone' and 'no' kids status for 'general' credit purposes have negative effect for repayment. 

3- 'alone' and 'yes' kids status in 'middle' and 'low' income level for 'real_estate' purpose are also risky. (9.4% - 9.9%)

4- 'real_estates' credit type for 'high' income level has the lowest for people who don't have kids and in 'together' category. (5.1%)

## Answering the questions

### Is there a relation between having kids and repaying a loan on time?

Yes, there is a relation between them. Having kids affects negatively repaying a loan on time.

#### Conclusion

There are about 14.000 people who don't have kids, it is almost 2 out of 3. Their debt rate is lower than average. On the contrary, people who have kids less reliable on repayment of loan. While their amount is about 7.000 (almost 1 out of 3), their debt rate is more than 1% of average debt rate. It is about 80 people. But if we compare it to those who do not have children, numbers increase and then it makes sense. It shows us having kids has negative effect on repayment of loan.

### Is there a relation between marital status and repaying a loan on time?

Yes, there is a relation between them.

#### Conclusion

The debt rate of families living together (8.0%) is almost same with the average (8.1%), while the debt rate of those living alone (8.5%) is above the general average. It can only slightly reduce the credit score of the alone customers.

If we look only from marital status, we see lonely people's high debt rate (8.5%). But if we look from another point to childless lonely people, we see that their debt rate (8.2%) is normal. It is said that lonely people without kids don't have a big problem about loan repayment because their debt rate (8.2%) is almost same general average (8.1%).

### Is there a relation between income level and repaying a loan on time?

Yes, there is a relation between these variables. We can say that high income people repays a loan on time accordings to others.

#### Conclusion

High income people have debt rate (6.9%) smaller than total average (8.1%). Low and medium-income people have same debt rate (8.3%). This situation is shows us 'income_level' affect total debt. In my opinion, it can be considered pointless for low or middle income level, but the credit score of high income levels can be increased.

### How do different loan purposes affect on-time repayment of the loan?

We categorized to 'general' and 'real_estate' topics. It affects and increases absolutely debt rates for 'general' category.

#### Conclusion

The 'general' credit type definitely has a negative effect, because it pulls all rates up on each income class.  Additionally it's obvious that debt rates for 'real_estate' are lower for each income class, so this credit type has positive effect for repayment, while 'general' has a negative.

## General conclusion

### Pivot Tables Summary

If we need to evaluate according to the average debt ratio (8.1%)

1- 1- Those who are in the 'middle' income and have kids have generally over average debt rates. They are the most risky group. (middle-with kids)

2- In 'middle' income level - for 'general' loan type, lonely people who have kids have highest debt rate. (12.6%)

2- 'alone' and 'no' kids status for 'general' credit purposes have negative effect for repayment. Their debt rates also generally over average.

3- 'alone' and 'yes' kids status in 'middle' and 'low' income level for 'real_estate' purpose are also risky. (9.4% - 9.9%)

4- 'real_estates' credit type for 'high' income level has the lowest for people who don't have kids and in 'together' category. (5.1%)



### General Summary


1- Lonely people have more difficulty with loan repayments than together.

2- Those who have children have more difficulty in repaying loans than those who do not.

3- Income levels of middle and low classes have no direct effect on loan repayment, but higher-income people appear to be more successful for repayments.

4- Loan type seems to have effect in repayment rates. The reason for this may be that the mortgage status in the real estate obliges the payment.

********